### PR TITLE
An attempt to implement AD for DG Kernels

### PR DESCRIPTION
**Pull request by fdkong**
Starting working on AD for DG kernels

### CODE REVIEW

**Comment by fdkong** [comment]
This now depends on #12738.

AD is working for DG Kernels now. But I would like to redesign a little bit so that we can use `computeQpResidual` instead of `computeADQpResidual` in users' code.

Create a DGKernelBase :)

**Comment by fdkong** [comment]
Now, `DGKernels` are using `computeQpResidual` instead of `computeADQpResidual`. This change makes `DGKernels` consistent with anywhere else.

**Comment by fdkong** [comment]
Will be going to fix all tests and add documentations.

**Comment by fdkong** [comment]
Should fix the `distributed mesh` test

**Comment by fdkong** [comment]
@lindsayad Do you think any more changes I need to make in this PR? We definitely can have more optimizations if we want.

File: framework/doc/content/source/actions/AddADDGKernelAction.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# AddADDGKernelAction
```
Comment (lindsayad) [inline framework/doc/content/source/actions/AddADDGKernelAction.md]: I thought we talked about not having a separate action and instead performing casts to figure out the correct warehouse for depositing things? I know this isn't currently what is being done for the other AD object, but we need to switch those over

Comment (fdkong) [reply framework/doc/content/source/actions/AddADDGKernelAction.md]: OK, I am going to make this work with the old `DGKernel` block.

**Comment by lindsayad** [comment]
Looks like we're still waiting on removal of `ADDGKernelAction`. And yes I know I need to get rid of the other `ADActions*`...

File: framework/include/materials/TwoMaterialPropertyInterface.h [hunk 2]
```diff
@@ -19,6 +19,8 @@ class TwoMaterialPropertyInterface;
 template <>
 InputParameters validParams<TwoMaterialPropertyInterface>();
 
+#define adGetNeighborMaterialProperty this->template getNeighborMaterialProperty
+
```
Comment (lindsayad) [inline framework/include/materials/TwoMaterialPropertyInterface.h]: I would like to see a `getNeighborADMaterialProperty` method added to this class with a corresponding `adGetNeighborADMaterialProperty` macro. The code should look something like:
```c++
template <typename T>
const ADMaterialPropertyObject<T> &
TwoMaterialPropertyInterface::getNeighborADMaterialProperty(const std::string & name)
{
  // Check if the supplied parameter is a valid input parameter key
  std::string prop_name = deducePropertyName(name);

  // Check if it's just a constant
  const ADMaterialPropertyObject<T> * default_property = defaultADMaterialProperty<T>(prop_name);
  if (default_property)
    return *default_property;
  else
    return _neighbor_material_data->getADProperty<T>(prop_name);
}
```
I would also really like to see a test with material properties that depend on a coupled variable to make sure that our ADDG system will work with ADMaterials

Comment (lindsayad) [reply framework/include/materials/TwoMaterialPropertyInterface.h]: I'll do this myself

File: framework/src/dgkernels/ADDGDiffusion.C [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+                        "diff",
+                        1.,
+                        "The diffusion (or thermal conductivity or viscosity) coefficient."););
+
+template <ComputeStage compute_stage>
+ADDGDiffusion<compute_stage>::ADDGDiffusion(const InputParameters & parameters)
+  : ADDGKernel<compute_stage>(parameters),
+    _epsilon(adGetParam<Real>("epsilon")),
+    _sigma(adGetParam<Real>("sigma")),
+    _diff(adGetMaterialProperty<Real>("diff")),
```
Comment (lindsayad) [inline framework/src/dgkernels/ADDGDiffusion.C]: Per recommendation in header, let's change to `adGetADMaterialProperty`

File: framework/src/dgkernels/ADDGKernel.C [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+template <>
+void
+ADDGKernel<RESIDUAL>::computeOffDiagJacobian(unsigned int)
+{
+}
+
+template <ComputeStage compute_stage>
+void
+ADDGKernel<compute_stage>::computeOffDiagElemNeighJacobian(Moose::DGJacobianType type,
```
Comment (lindsayad) [inline framework/src/dgkernels/ADDGKernel.C]: It looks like you've chosen not to optimize like I have for `ADKernels` where I've eliminated the outer coupled variable loop. I think that's fine for now, but we should make that optimization in the not too distant future

File: framework/src/dgkernels/ADDGKernel.C [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+          (type == Moose::ElementElement || type == Moose::ElementNeighbor) ? Moose::Element
+                                                                            : Moose::Neighbor);
+      for (_j = 0; _j < loc_phi.size(); _j++)
+        _local_ke(_i, _j) += _JxW[_qp] * _coord[_qp] * residual.derivatives()[ad_offset + _j];
+    }
+
+  accumulateTaggedLocalMatrix();
+}
+
+template class ADDGKernel<RESIDUAL>;
```
Comment (lindsayad) [inline framework/src/dgkernels/ADDGKernel.C]: You can use the macro that @dschwen created `adBaseClass(ADDGKernel)` which automatically does these two lines

File: test/tests/dgkernels/ad_dg_diffusion/tests [hunk 6]
```diff
@@ -0,0 +1,10 @@
+[Tests]
+  [./test]
+    type = 'Exodiff'
+    input = '2d_diffusion_ad_dg_test.i'
+    exodiff = '2d_diffusion_ad_dg_test_out.e-s003'
```
Comment (lindsayad) [inline test/tests/dgkernels/ad_dg_diffusion/tests]: Why the `-s003` extension? We're not doing adaptivity right? Seems like we only need the `2d_diffusion_ad_dg_test_out.e` binary...?

**Review by lindsayad — CHANGES_REQUESTED** [review]
This is a really neat PR. Sorry I haven't reviewed sooner. See comments

File: test/tests/dgkernels/ad_dg_diffusion/tests (hunk 6 above)
Comment (lindsayad) [reply test/tests/dgkernels/ad_dg_diffusion/tests]: Ah, I see the `Adaptivity` block now. let's test both files

File: framework/doc/content/source/dgkernels/ADDGDiffusion.md [hunk 7]
```diff
@@ -0,0 +1,3 @@
+# ADDGDiffusion<JACOBIAN>
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/doc/content/source/dgkernels/ADDGDiffusion.md]: At the end of the file, let's add:
```
!syntax parameters /DGKernels/ADDGDiffusion<RESIDUAL>

!syntax inputs /DGKernels/ADDGDiffusion<RESIDUAL>

!syntax children /DGKernels/ADDGDiffusion<RESIDUAL>
```

**Comment by fdkong** [comment]
@lindsayad I will address the comments ASAP.

**Comment by lindsayad** [comment]
@tophmatthews see https://github.com/idaholab/moose/pull/12743/commits/682fab9a52686261250f50e53d57f3d09e9c6ead

**Review by lindsayad — APPROVED** [review]

**Review by aeslaughter — APPROVED** [review]

**Pull request merged**